Plugins/NativeEngine/Source/VertexBuffer.h

```diff
@@ -16,41 +17,94 @@ namespace Babylon
+
+        void Set(bgfx::Encoder* encoder, uint8_t& streamCount, uint32_t startVertex, uint32_t numVertices);
+
+        struct InstanceInfo
         {
             VertexBuffer* Buffer{};
             uint32_t Offset{};
             uint32_t Stride{};
             uint32_t ElementSize{};
         };
 
-        static void BuildInstanceDataBuffer(bgfx::InstanceDataBuffer& instanceDataBuffer, const std::map<bgfx::Attrib::Enum, InstanceVertexBufferRecord>& vertexBufferInstance, uint32_t instanceCount);
+        static void BuildInstanceDataBuffer(bgfx::InstanceDataBuffer& instanceDataBuffer, const std::map<bgfx::Attrib::Enum, InstanceInfo>& instances, uint32_t instanceCount);
 
     private:
-        std::optional<std::vector<uint8_t>> m_bytes{};
-        bool m_dynamic{};
+        void Build(uint32_t numVertices);
 
-        union
+        class Handle final
         {
-            bgfx::VertexBufferHandle m_handle{bgfx::kInvalidHandle};
-            bgfx::DynamicVertexBufferHandle m_dynamicHandle;
+        public:
+            Handle(Graphics::DeviceContext& deviceContext, std::vector<uint8_t> bytes, bool dynamic, uint16_t byteStride);
+            ~Handle();
+
+            // Copy semantics
+            Handle(const Handle&) = delete;
+            Handle& operator=(const Handle&) = delete;
+
+            // Move semantics
+            Handle(Handle&&) noexcept;
+            Handle& operator=(Handle&&) noexcept;
+
+            void Update(gsl::span<uint8_t> bytes, uint32_t startVertex);
+
+            void Set(bgfx::Encoder* encoder, uint8_t stream, uint32_t startVertex, uint32_t numVertices, bgfx::VertexLayoutHandle layoutHandle);
+
+        private:
+            Graphics::DeviceContext& m_deviceContext;
+            const uintptr_t m_deviceId{};
+            bool m_dynamic{};
+
+            union
+            {
+                bgfx::VertexBufferHandle m_handle{bgfx::kInvalidHandle};
+                bgfx::DynamicVertexBufferHandle m_dynamicHandle;
+            };
         };
 
-        bool m_disposed{};
-        uintptr_t m_deviceID;
+    private:
         Graphics::DeviceContext& m_deviceContext;
+
+        std::vector<uint8_t> m_bytes{};
+        const bool m_dynamic{};
+        bool m_buildCalled{};
+
+        struct AttributeInfo
+        {
+            bgfx::AttribType::Enum AttribType{};
+            uint32_t ByteOffset{};
+            uint8_t NumElements{};
+            bool Normalized{};
+        };
+
+        uint16_t m_byteStride{};
+        std::map<bgfx::Attrib::Enum, AttributeInfo> m_attributes{};
+
+        std::optional<Handle> m_handle{};
+
+        struct StreamInfo
+        {
+            const AttributeInfo& Attribute;
+            std::optional<Handle> PromoteToFloatsHandle;
+            uint32_t Offset;
+            bgfx::VertexLayoutHandle LayoutHandle;
+        };
+
+        std::list<StreamInfo> m_streams{};
```

Comment:
Why a list instead of a vector?

Reply:
I was trying to avoid adding move semantics, but it's probably not necessary.